### ISSUE TITLE
Possible pool resource release bug

### ISSUE THREAD

**Issue by andreisilviudragnea** [reactor/reactor-pool#124]
There are times when the resources are never released back to the pool. I do not know the cause yet, but I was able to reproduce the problem in a separate project [here](https://github.com/andreisilviudragnea/project-reactor-bugs/blob/master/src/test/java/com/example/ReactorPoolBugTest.java).

## Expected Behavior

The resources should be released to the pool.
## Actual Behavior

The resources never get released to the pool.
## Steps to Reproduce

```java
package com.example;

import org.junit.jupiter.api.Test;

import java.time.Duration;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.pool.InstrumentedPool;
import reactor.pool.PoolBuilder;

public class ReactorPoolBugTest {

    private static final int COUNT = 10_000;

    private final InstrumentedPool<String> stringReactivePool = PoolBuilder
            .from(Mono.just("value").delayElement(Duration.ofMillis(2)))
            .maxPendingAcquireUnbounded()
            .sizeBetween(0, 3)
            .buildPool();

    @Test
    public void reactorPoolBug() throws InterruptedException {
        ExecutorService loggerThreads = Executors.newFixedThreadPool(
                1,
                r -> {
                    Thread t = Executors.defaultThreadFactory().newThread(r);
                    t.setDaemon(true);
                    return t;
                }
        );

        loggerThreads.submit(new PoolMetricsLogger(stringReactivePool.metrics()));

        ExecutorService executorService = Executors.newFixedThreadPool(
                16,
                r -> {
                    Thread t = Executors.defaultThreadFactory().newThread(r);
                    t.setDaemon(true);
                    return t;
                }
        );

        CountDownLatch cdl = new CountDownLatch(COUNT);
        for (int i = 0; i < COUNT; i++) {
            executorService.submit(new FlatMapErrorTask(cdl));
        }

        cdl.await();
    }

    private static final class PoolMetricsLogger implements Runnable {

        private final InstrumentedPool.PoolMetrics poolMetrics;

        private PoolMetricsLogger(InstrumentedPool.PoolMetrics poolMetrics) {
            this.poolMetrics = poolMetrics;
        }

        public void run() {
            while (true) {
                try {
                    Thread.sleep(200);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }

                System.err.printf(
                        "[POOL Metrics] Acquired = %d Pending = %d Idle = %d%n",
                        poolMetrics.acquiredSize(),
                        poolMetrics.pendingAcquireSize(),
                        poolMetrics.idleSize()
                );
            }
        }
    }

    private final class FlatMapErrorTask implements Runnable {

        private final CountDownLatch cdl;

        public FlatMapErrorTask(CountDownLatch cdl) {
            this.cdl = cdl;
        }

        public void run() {
            Flux<Void> flux = Flux
                    .range(0, 10)
                    .flatMap(i -> stringReactivePool
                            .withPoolable(value -> Mono
                                    .just(value)
                                    .delayElement(Duration.ofMillis(10))
                                    .then()
                            )
                            .switchIfEmpty(Mono.error(new RuntimeException("Empty")))
                    )
                    .doOnComplete(() -> cdl.countDown())
                    .doOnError(error -> cdl.countDown());

            try {
                flux.blockLast();
            } catch (Exception e) {
                System.err.println(e);
            }

            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}
```

## Possible Solution
The test should always finish, but most of the time it never does, because all the pool resources remain acquired. For an unknown reason, the test fails way less often if you remove the `Thread.sleep()` at the end.

## Your Environment

* Reactor version(s) used: `reactor-core:3.4.4`, `reactor-pool:0.2.3`
* Other relevant libraries versions (eg. `netty`, ...): 
* JVM version (`java -version`):
```
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (Zulu 8.52.0.23-CA-linux64) (build 1.8.0_282-b08)
OpenJDK 64-Bit Server VM (Zulu 8.52.0.23-CA-linux64) (build 25.282-b08, mixed mode)
```
* OS and version (eg `uname -a`): `Linux andrei-desktop 5.8.0-44-generic #50-Ubuntu SMP Tue Feb 9 06:29:41 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by simonbasle** [comment]
just in case this is a bug with the `withPoolable` construct, have you been able to reproduce with the `acquire` API?

**Comment by andreisilviudragnea** [comment]
@simonbasle I tried reproducing the bug only on the pool side [here](https://github.com/andreisilviudragnea/project-reactor-bugs/blob/master/src/test/java/com/example/ReactorPoolIsolatedBugTest.java) and only on the `Flux.usingWhen()` operator side [here](https://github.com/andreisilviudragnea/project-reactor-bugs/blob/master/src/test/java/com/example/UsingWhenBugTest.java), but I failed.

I am not very familiar with the inner workings of either `reactor-core` or `reactor-pool` projects, but I may not use the `.doOnCancel()` operator correctly in the first isolated example above.

**Comment by andreisilviudragnea** [comment]
@simonbasle I have opened a `MonoUsingWhen` / `FluxUsingWhen` issue https://github.com/reactor/reactor-core/issues/2661

**Comment by andreisilviudragnea** [comment]
@simonbasle After investigating the reactor-pool implementation more, should `drainLoop()` be called also when `AbstractPool.cancelAcquire()` is called?

**Comment by andreisilviudragnea** [comment]
It looks like activating background eviction makes the test fail less often, but the test for the bug still fails sometimes

**Comment by andreisilviudragnea** [comment]
I managed to isolate the reactor-pool bug [here](https://github.com/andreisilviudragnea/project-reactor-bugs/blob/master/src/test/java/com/example/ReactorPoolIsolatedBugTest.java).

**Comment by agorbachenko** [comment]
Hi there! According to @mp911de and @chemicL this issue is the same as https://github.com/r2dbc/r2dbc-pool/issues/198. That issue is another reproducer of this critical bug. Maybe this helps to spot the bug and fix it.

**Comment by chemicL** [comment]
@andreisilviudragnea thanks for the report. I'm sorry to see it didn't get much attention. With some other reports around r2dbc, jOOQ, and reactor-core I stumbled upon the same problem and discovered your report. So I'll use this issue to try to make some progress on resolving this.

Here's a highly reproducible example that I created to capture what I observed in the r2dbc-pool issue mentioned by @agorbachenko (best to run it in a loop / repeat mode in the IDE): https://github.com/reactor/reactor-pool/compare/fix-124-deliver-cancel-race?expand=1.

I'll chat with the team about possible solutions.
